File: java/src/main/java/ch/frostnova/example/shoppingcart/model/CurrencyAmount.java
package ch.frostnova.example.shoppingcart.model;

import java.math.BigDecimal;
import java.util.Objects;

import static ch.frostnova.example.shoppingcart.util.Check.format;
import static ch.frostnova.example.shoppingcart.util.Check.min;
import static ch.frostnova.example.shoppingcart.util.Check.required;
import static java.math.RoundingMode.HALF_UP;

public class CurrencyAmount {

    private final String currency;
    private final BigDecimal amount;

    public CurrencyAmount(String currency, BigDecimal amount) {
        this.currency = required("currency", currency, format("[A-Z]{3}", "must be a currency code"));
        this.amount = required("amount", amount, min(0)).setScale(2, HALF_UP);
    }

    public CurrencyAmount(String currency, Number amount) {
        this(currency, new BigDecimal(required("amount", amount).toString()));
    }

    public String getCurrency() {
        return currency;
    }

    public BigDecimal getAmount() {
        return amount;
    }

    @Override
    public String toString() {
        return String.format("%s %s", currency, amount);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        CurrencyAmount that = (CurrencyAmount) o;
        return Objects.equals(currency, that.currency) &&
                Objects.equals(amount, that.amount);
    }

    @Override
    public int hashCode() {
        return Objects.hash(currency, amount);
    }
}

File: kotlin/src/main/kotlin/ch/frostnova/example/shoppingcart/model/CurrencyAmount.kt
package ch.frostnova.example.shoppingcart.model

import ch.frostnova.example.shoppingcart.util.Check.format
import ch.frostnova.example.shoppingcart.util.Check.min
import ch.frostnova.example.shoppingcart.util.Check.required
import java.math.BigDecimal
import java.math.RoundingMode.HALF_UP
import java.util.*

class CurrencyAmount(currency: String, amount: BigDecimal, test: String = "Hello") {

    val currency: String = required("currency", currency, format("[A-Z]{3}", "must be a currency code"))
    val amount: BigDecimal = required("amount", amount, min(0)).setScale(2, HALF_UP)

    constructor(currency: String, amount: Number) : this(currency, BigDecimal(required<Number>("amount", amount).toString()))

    override fun toString(): String {
        return "$currency $amount"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as CurrencyAmount

        if (currency != other.currency) return false
        if (amount != other.amount) return false

        return true
    }

    override fun hashCode(): Int {
        return Objects.hash(currency, amount)
    }
}